Rithika taking over for Kanishk: Phase 4 - Task Comments Frontend Implementation

kanishkagarwal6101 opened the pull request:
# Description
Implements Phase 4 - Task Comments Frontend feature for the Education Portal. This PR adds comprehensive comment functionality to the WriteTaskUpload component, allowing students to post comments and educators to view them, with full dark mode support and delete functionality.

## Related PRS (if any):
This is a standalone frontend feature. No related backend PRs at this time.

## Main changes explained:
- **Create file CommentBox.jsx** - Comment input component with validation and character counter
- **Create file CommentList.jsx** - Comment display component with delete functionality for own comments
- **Create file Toast.jsx** - Toast notification system for success/error feedback
- **Create file useToast.js** - Custom hook for managing toast notifications
- **Create file CommentBox.module.css** - Styling for comment input with dark mode support
- **Create file CommentList.module.css** - Styling for comment list with dark mode support
- **Create file Toast.module.css** - Styling for toast notifications
- **Update file WriteTaskUpload.jsx** - Integrated comment components with localStorage persistence
- **Update file WriteTaskUpload.module.css** - Added comprehensive dark mode styling for entire component

## How to test:
1. Check into current branch: `git checkout kanishk_task_comments_frontend`
2. Run `npm install` to install dependencies
3. Run the app locally: `npm run start:local`
4. Clear site data/cache
5. Navigate to: `http://localhost:3000/educationportal/tasks/upload`
6. **Test comment submission:**
   - Type a comment in the "Comments/Queries Section"
   - Click "Submit" - should show success toast
   - Comment should appear in the list below
7. **Test validation:**
   - Try submitting empty comment - should show inline error
   - Try submitting 1000+ characters - should show character limit error
8. **Test delete functionality:**
   - Click the red "×" button on your own comments
   - Should show success toast and remove comment
9. **Test dark mode:**
   - Toggle dark mode - all icons, text, and components should be properly styled
   - Comments should persist after page refresh

## Screenshots or videos of changes:
- Comment input with validation and character counter
- Comment list with delete buttons for own comments
- Toast notifications for success/error feedback
- Full dark mode support across all components
- Responsive design with proper spacing

## Note:
- Comments are persisted using localStorage (task-specific storage)
- Delete functionality only available for comments posted by current user
- All components follow the existing design system and dark mode patterns
- Ready for backend integration - API endpoints can be easily connected when available
- No breaking changes to existing functionality

kanishkagarwal6101 committed:
Add comment functionality with dark mode support and delete feature

kanishkagarwal6101 committed:
Add comment functionality with dark mode support and delete feature

Anusha-Gali reviewed (changes requested):
Hi Kanishk,

I have reviewed your PR locally - i only have 2 concerns.
1) **Try submitting 1000+ characters - should show character limit error** You mentioned character limit error - it only stops me due to character limit inline and doesn't throw any error - is it supposed to inline as if it not then i do not see any error being displayed.
2) **Delete functionality only available for comments posted by current user** - I am able to delete someone else's comment as you can see in the below video. I have left my comments as is and have also tested this using other login's and was able to delete comments left at the end of below video.
https://github.com/user-attachments/assets/1350b613-4294-451c-85bf-aecd4c7aa329

kanishkagarwal6101 committed:
Fix comment validation and delete security issues

- Fix character limit error: Remove maxLength attribute and show inline error when exceeding 1000 characters
- Fix empty comment error: Enable submit button to allow error display on submit
- Fix delete security: Use userId comparison instead of name comparison to prevent unauthorized deletions
- Add visual feedback: Character count turns red when exceeding limit
- Add server-side validation: Double-check userId in handleDeleteComment for additional security

kanishkagarwal6101 committed:
Fix comment validation and delete security issues

- Fix character limit error: Remove maxLength attribute and show inline error when exceeding 1000 characters
- Fix empty comment error: Enable submit button to allow error display on submit
- Fix delete security: Use userId comparison instead of name comparison to prevent unauthorized deletions
- Add visual feedback: Character count turns red when exceeding limit
- Add server-side validation: Double-check userId in handleDeleteComment for additional security

kanishkagarwal6101 committed:
Fix SonarQube code quality issues

- Add PropTypes validation to all components (CommentBox, CommentList, Toast, Icon, BellIcon)
- Improve error handling with development-mode logging
- Fix CSS contrast issues (improved text colors for better accessibility)
- Remove duplicate CSS selectors (:root and .headerDark)
- Extract validation logic to reduce cognitive complexity
- Improve code maintainability and reliability

kanishkagarwal6101 committed:
Fix SonarQube code quality issues

- Add PropTypes validation to all components (CommentBox, CommentList, Toast, Icon, BellIcon)
- Improve error handling with development-mode logging
- Fix CSS contrast issues (improved text colors for better accessibility)
- Remove duplicate CSS selectors (:root and .headerDark)
- Extract validation logic to reduce cognitive complexity
- Improve code maintainability and reliability

kanishkagarwal6101 committed:
fix: resolve SonarQube issues - improve contrast and reduce cognitive complexity

- Fix contrast issue in CommentBox.module.css by using darker blue (#2563eb) for better WCAG compliance
- Reduce cognitive complexity in WriteTaskUpload.jsx by extracting render helpers and event handlers
- Extract functions: renderSidebar, renderHeader, renderFormSection, renderHistorySection, renderProgressPane, renderFormGrid
- Extract event handlers: handleCommentChange, handleLinkUrlChange, handleShowLink
- Extract data management: loadCommentsFromStorage, getDefaultComments, saveCommentsToStorage, createNewComment, formatItemTime
- Extract validation: validateDeletePermission

kanishkagarwal6101 committed:
fix: resolve SonarQube issues - improve contrast and reduce cognitive complexity

- Fix contrast issue in CommentBox.module.css by using darker blue (#2563eb) for better WCAG compliance
- Reduce cognitive complexity in WriteTaskUpload.jsx by extracting render helpers and event handlers
- Extract functions: renderSidebar, renderHeader, renderFormSection, renderHistorySection, renderProgressPane, renderFormGrid
- Extract event handlers: handleCommentChange, handleLinkUrlChange, handleShowLink
- Extract data management: loadCommentsFromStorage, getDefaultComments, saveCommentsToStorage, createNewComment, formatItemTime
- Extract validation: validateDeletePermission

kanishkagarwal6101 committed:
fix: improve contrast in CommentList avatar for WCAG compliance

- Change color from 'white' to '#ffffff' for better specificity
- Use darker gradient colors (#2563eb, #7c3aed) for better contrast
- Update dark mode gradient to even darker shades (#1e40af, #6d28d9)

kanishkagarwal6101 committed:
fix: improve contrast in CommentList avatar for WCAG compliance

- Change color from 'white' to '#ffffff' for better specificity
- Use darker gradient colors (#2563eb, #7c3aed) for better contrast
- Update dark mode gradient to even darker shades (#1e40af, #6d28d9)

kanishkagarwal6101 committed:
fix: replace Math.random() with crypto.randomUUID() for secure ID generation

- Replace Date.now() + Math.random() with crypto.randomUUID() in useToast hook
- Use cryptographically secure random number generator to satisfy SonarQube security hotspot
- crypto.randomUUID() is available in all modern browsers and provides proper UUIDs

kanishkagarwal6101 committed:
fix: replace Math.random() with crypto.randomUUID() for secure ID generation

- Replace Date.now() + Math.random() with crypto.randomUUID() in useToast hook
- Use cryptographically secure random number generator to satisfy SonarQube security hotspot
- crypto.randomUUID() is available in all modern browsers and provides proper UUIDs

debadyuti23 reviewed (approved):
Hi @kanishkagarwal6101 I have verified the changes, it works correctly. Couple of observations which I can see:
1. The submit button and pie chart got removed when I submitted a very long comment.
2. In dark mode, various functionalities like the progress bar, pie chart does not adjust with proper color.

Test video 1: https://drive.google.com/file/d/1SqTdH3_agCdtP8EZY3ffmN9Uyw_LyCWu/view?usp=sharing

Test video 2: https://github.com/user-attachments/assets/6299a3e1-8eba-48d5-89f3-fa0df0747e78



Since these changes are not related to PR, I have approved it. However please be mindful about these issues in future.

rohanrastogi311 reviewed (changes requested):
Functionality works well, well done. Need to do dark mode fixes. Lot's of white boxes still showing, check image below. 

<img width="1722" height="1035" alt="PR 4227 Screenshot" src="https://github.com/user-attachments/assets/04b894ec-3259-4baf-adfd-8213af815bfc" />
<img width="1728" height="924" alt="PR 4227 Screenshot 2" src="https://github.com/user-attachments/assets/43935eae-dc9a-4ee6-9d17-bce8158e29d9" />
<img width="1728" height="1041" alt="PR 4227 Screenshot 3" src="https://github.com/user-attachments/assets/642f36dc-c5e3-4dc1-8985-4f73a366a580" />

ShreyaMP1999 reviewed (approved):
Tested the current branch locally following the provided steps. Comment submission, validation (empty and character limit), delete functionality, dark mode styling, and comment persistence after refresh are all working as expected. No issues found.


https://github.com/user-attachments/assets/e1225edb-3a25-4972-8f9b-034bc78e1cf0

Anusha-Gali reviewed (approved):
Hi Kanishk,

I have re-reviewed your PR and thanks for implementing the change i had suggested. All requirements works as expected
<img width="1271" height="757" alt="Screenshot 2026-01-06 at 8 12 56 PM" src="https://github.com/user-attachments/assets/2f22b4f9-1491-442d-89c7-c8387d461d92" />
<img width="1271" height="757" alt="Screenshot 2026-01-06 at 8 13 17 PM" src="https://github.com/user-attachments/assets/fa90a7da-55fa-4e48-b034-f11b40aa4b64" />
<img width="1271" height="757" alt="Screenshot 2026-01-06 at 8 14 00 PM" src="https://github.com/user-attachments/assets/485a62f2-f527-49bd-81d6-003518740b3a" />
<img width="1271" height="757" alt="Screenshot 2026-01-06 at 8 14 31 PM" src="https://github.com/user-attachments/assets/fa2819f6-8761-4ae6-8537-d9bef9bcc503" />
<img width="1271" height="757" alt="Screenshot 2026-01-06 at 8 14 59 PM" src="https://github.com/user-attachments/assets/7586b238-0dff-4c33-b32e-5aa956e39a40" />

Shravan-neelamsetty reviewed (approved):
Hi Kanishk, I tested this PR locally by navigating to /educationportal/tasks/upload and verified the comment functionality. Comment submission works correctly with proper validation for empty comments and character limits. The character counter displays accurately, and comments persist after page refresh as expected. The delete functionality correctly restricts deletion to only the current user's comments. However, I noticed dark mode issues: there are several white boxes visible in dark mode, and some elements like the progress bar and pie chart don't adjust with proper dark mode colors.

@
<img width="1466" height="872" alt="Screenshot 2026-01-08 at 6 05 48 PM" src="https://github.com/user-attachments/assets/cdc9e912-da68-4803-bfc9-f7c7d0ac0e36" />
<img width="1067" height="259" alt="Screenshot 2026-01-08 at 6 06 02 PM" src="https://github.com/user-attachments/assets/a69aaf46-7b57-4136-a8a0-689f87e17a22" />
<img width="1470" height="818" alt="Screenshot 2026-01-08 at 6 06 31 PM" src="https://github.com/user-attachments/assets/0a6d3823-ad9d-4fda-950e-fd18f1019f83" />

Vikas-8055 reviewed (commented):
Hi Kanishk,

I have reviewed your PR locally and tested the comment functionality. The main features are working correctly - comment submission with character counter, validation for empty comments and 1000+ character limit, delete functionality for own comments, and comment persistence after refresh all work as expected. I tested in both light and dark modes, and the Comments/Queries Section displays properly with readable text and buttons in both modes.

I noticed some white boxes visible in dark mode (Unit 1 selection button, Choose Trades button, and file upload area), but as other reviewers mentioned, these appear to be part of the existing page layout and outside the scope of this PR which specifically focuses on the comment functionality. The comment components themselves have proper dark mode styling.

<img width="1512" height="856" alt="Screenshot 2026-01-08 at 6 22 32 PM" src="https://github.com/user-attachments/assets/da29de95-807b-4239-8712-7f6ffa571e3e" />

<img width="1512" height="855" alt="Screenshot 2026-01-08 at 6 22 45 PM" src="https://github.com/user-attachments/assets/3c2cc24d-ebce-4505-9808-a777e6335fdc" />

Ganesh112001 reviewed (approved):
Hi Kanishk, I tested PR #4227 and verified that the Task Comments feature works perfectly. After navigating to /educationportal/tasks/upload, I confirmed that students can post comments with proper validation and character counter, delete their own comments with the red × button, and receive success/error toast notifications. Dark mode is fully supported with excellent contrast across all components. Comments persist correctly using localStorage after page refresh. No console errors observed. Everything functions perfectly and is ready for merge!

<img width="1470" height="822" alt="Screenshot 2026-01-18 at 1 20 26 AM" src="https://github.com/user-attachments/assets/2004351a-a9f8-41cf-844c-93d0db604509" />

<img width="1470" height="829" alt="Screenshot 2026-01-18 at 1 20 46 AM" src="https://github.com/user-attachments/assets/3bc5a3bd-a719-4ef4-b1a1-4d8ac7ab950f" />

Vikas-8055 reviewed (approved):
Hi Kanishk,

I reviewed your PR locally and all core features work great! Comment submission works smoothly with the character counter, empty comment validation shows inline errors, and the 1000+ character limit displays the proper error message. Delete functionality works correctly I can only delete my own comments and the success toast appears as expected. Comments persist after page refresh.

Regarding @rohanrastogi311's dark mode feedback the Progress Bar and Drag & Drop areas are pre-existing components outside this PR's scope. A separate PR should be raised to address those dark mode issues. This PR focuses on the Task Comments feature, and the Comments/Queries Section is properly styled in both light and dark mode.




<img width="1512" height="854" alt="Screenshot 2026-01-30 at 9 19 08 PM" src="https://github.com/user-attachments/assets/82b2df85-1ac3-4e70-b25d-36961d03904e" />
<img width="1509" height="858" alt="Screenshot 2026-01-30 at 9 18 55 PM" src="https://github.com/user-attachments/assets/f555bf18-8d3b-4d49-96ed-3849172527d0" />
<img width="1508" height="853" alt="Screenshot 2026-01-30 at 9 18 47 PM" src="https://github.com/user-attachments/assets/8ecec3b6-9301-4d1d-a968-ee760803dff6" />

roshini1212 reviewed (approved):
Tested the kanishk_task_comments_frontend branch locally and all functionality works perfectly at "/educationportal/tasks/upload". Comment submission displays success toasts and lists correctly, validation properly blocks empty submissions and 1000+ character inputs with clear inline errors, delete removes own comments with confirmation toasts and light mode styling is fully responsive across all components. 
The dark mode has some white boxes which felt off. But the basic functionalities according to the PR are working properly. 

<img width="1500" height="840" alt="Image" src="https://github.com/user-attachments/assets/eb9494db-55f4-4894-a156-bc8cb5fc34c4" />

<img width="1500" height="833" alt="Image" src="https://github.com/user-attachments/assets/0361128c-1958-4d0d-b42b-ca1c822834f6" />

<img width="1491" height="807" alt="Image" src="https://github.com/user-attachments/assets/b4000374-edcc-4d8d-8145-21b498457f29" />

naznin07 reviewed (approved):
Controlled textarea with validation.
<img width="1772" height="566" alt="PR #4227" src="https://github.com/user-attachments/assets/daa2b75b-aa9a-4c97-b590-d3ee81effb6a" />

ManojPuttaswamy reviewed (changes requested):
Tested all functionality and everything works as expected. One minor fix needed — the submit button should be disabled when the comment exceeds 1000 characters. This can be done by updating the disabled prop on the button. 

disabled={disabled || comment.length > 1000}

saisandeepkoritala reviewed (commented):
Hi  Kaniskh,

I ran your PR locally and all seems good in light and dark mode, everything works as intended. But i was able to delete other user comment.

<img width="1544" height="965" alt="Image" src="https://github.com/user-attachments/assets/927394eb-c2cc-412b-92ea-a6e493c40fd9" />

SharadhaKasiviswanathan commented:
Tested this locally and reproduced the issue: when 'currentUserId' is missing, clicking Submit shows the error toast 'Unable to identify user. Please refresh the page and try again', but the text area is still cleared, so the user loses their draft even though nothing was posted. This comes from CommentBox clearing its local state immediately after calling onSubmit(), while handleCommentSubmit() in 'WriteTaskUpload' can early return on failure. Can the flow be updated so that the input is cleared only after the parent confirms a successful submit?
Also, another issue is that a user is able to delete all the other comments as well (attached in the last screenshot).
<img width="1440" height="2200" alt="image" src="https://github.com/user-attachments/assets/c52b9b03-a977-4101-b0e9-a491860aaa76" />
<img width="1440" height="2200" alt="image" src="https://github.com/user-attachments/assets/f95e512f-2d1c-47e1-a0e1-6293dfecbd4d" />
<img width="1902" height="850" alt="Screenshot 2026-03-29 at 5 59 48 AM" src="https://github.com/user-attachments/assets/eb26c341-676a-4b24-859b-3832a94dfbcb" />

rajanidi1999 reviewed (approved):
Overall looks good. The components are well-structured and reusable, and the feature implementation is solid. Just a few minor improvements to consider: remove unnecessary Promise.resolve() in async functions, standardize API response/error shapes for consistency, and disable the submit button in CommentBox when validation fails (e.g., character limit exceeded). These are non-blocking and can be addressed in follow-ups.

rajanidi1999 reviewed (approved):
Hi,
i have tested your PR locally and its working well.

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/7e070f7c-bd85-4183-b575-5def831c2116" />

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/81bf6089-a1f8-4c8e-a50e-c5cc7c6710a3" />

rithika-paii reviewed (changes requested):
Hi Kanishk,

I reviewed this PR and tested the Task Comments functionality on the upload page in both light and dark modes. The overall feature works as expected, but I found one issue:

Comment deletion permission issue: Currently, it is possible to delete comments posted by other users. The delete option should be restricted so that users can only delete their own comments, as mentioned in the PR description.

https://github.com/user-attachments/assets/17ef32c9-cc3a-465a-b4b6-d5c474824f91

saurabhdipte reviewed (changes requested):
Hi Vikas,

I tested the changes locally. I’m able to submit a comment and delete it successfully. However, I also noticed that I’m able to delete other students’ comments, which might need to be restricted.

The validation for empty comments is working as expected. For the 1000+ character limit, I’m able to enter more than 1000 characters in the text box, but submission is blocked. It might be better to restrict input beyond 1000 characters at the typing level itself.

In dark mode, there are a few contrast-related issues, particularly with some text boxes, which may need adjustment.

<img width="1896" height="935" alt="Image" src="https://github.com/user-attachments/assets/46afb50f-071d-4b0c-a392-694bdeedab8b" />

kunchalasireesha reviewed (changes requested):
Hi Vikas, it looks like this PR currently has merge conflicts with the base branch. Could you please rebase or merge the latest changes from development into your branch and resolve the conflicts? Once the build passes, I’ll finish the review.

DeepighaJ reviewed (dismissed):
Checked out to current branch and tested PR locally.
Navigate to /educationportal/tasks/upload
Tested comment submission and success toast works as expected:
Empty comment submission and 1000+ character shows inline error as expected
Clicked the red "×" button to remove current user and not available for other user comment and success toast works as expected.
Comments persist after page refresh as expected.
Chart layout looks large and misaligned. Approving since its out of scope for this PR.

<img width="1600" height="852" alt="Image" src="https://github.com/user-attachments/assets/3ae6e15c-24dc-46eb-bc88-d247c7167ae1" />
<img width="1600" height="852" alt="Image" src="https://github.com/user-attachments/assets/628b8b74-6f81-43e3-87dd-47562e35f1fa" />
<img width="1600" height="852" alt="Image" src="https://github.com/user-attachments/assets/1dac2402-d443-44d7-bce4-cd3df7290e92" />
<img width="1600" height="852" alt="Image" src="https://github.com/user-attachments/assets/292d9285-2e46-4ad0-8fa1-c8f614777928" />
<img width="1600" height="852" alt="Image" src="https://github.com/user-attachments/assets/8aa79b19-9a57-4f12-92ff-76ecdecd7681" />
<img width="1600" height="852" alt="Image" src="https://github.com/user-attachments/assets/c7bea133-53a3-4d08-b1a8-4046ff018a8f" />
<img width="1600" height="852" alt="Image" src="https://github.com/user-attachments/assets/22b4fee2-e274-427a-af5b-eeb622a922f5" />
<img width="1600" height="852" alt="Image" src="https://github.com/user-attachments/assets/38f334d9-9dc0-4771-96be-f6102be6ab17" />

rithika-paii committed:
fix: add missing dark mode classes for progress card, chart card, progress bar, and progress row

kzou55 reviewed (changes requested):
Hi Rithika, 

I ran and tested the PR locally:

## Verified
- User can type and submit a comment with the success toast appearing
- Inline error appears if there's an empty comment
- Inline error appears if there's a comment with 1000+ characters
- Users can delete a comment and a success toast will appear
- Functionality works in Dark Mode
- Comments persists after page refresh

Testing in Light Mode:

https://github.com/user-attachments/assets/e3792932-4ee9-4dc6-b212-051fcb6401ee

Testing in Dark Mode:

https://github.com/user-attachments/assets/b21967f0-84d8-42e2-a5a4-7b398afa4d82


Issues:
- In Dark Mode: Shouldn't the inline error messages be red?

    <img width="1272" height="609" alt="Image" src="https://github.com/user-attachments/assets/05a40a2e-7d5d-4ddd-839a-b0a0dd77d199" />

    <img width="1312" height="619" alt="Image" src="https://github.com/user-attachments/assets/b37035e5-4dd1-4763-8053-ae87cbe0b777" />

    <img width="1315" height="661" alt="Image" src="https://github.com/user-attachments/assets/16917933-9343-4e52-a0e0-c86a758074ce" />

DeepighaJ reviewed (changes requested):
- Tested the PR locally.
- The progress bar getting scrolled along with the page so there is a change is layout of Comment/Queries section which was not like this previously when i tested.
- As per previous comment the inline error message should be in red.
- the comment submission and success toast works as expected.

<img width="1600" height="852" alt="Image" src="https://github.com/user-attachments/assets/65d8f397-49c3-42b1-81e9-beff76fcf37a" />
<img width="1600" height="852" alt="Image" src="https://github.com/user-attachments/assets/5b95c2d2-65e5-4998-85d5-cc09a54704c5" />
<img width="1600" height="852" alt="Image" src="https://github.com/user-attachments/assets/ec8d5d14-f711-40ad-a744-63ad753b5f38" />

rithika-paii committed:
fix: error message color override in dark mode, remove sticky progress pane
